### PR TITLE
[HigherOrderOp] support while_loop in dynamo

### PR DESCRIPTION
**Pull request by ydwu4**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #116913

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/_higher_order_ops/while_loop.py [hunk 1]
```diff
@@ -97,8 +97,27 @@ def body_fn(iter, x):
         - 'while_loop' only supports **inference** right now. Autograd will be supported in the future.
 
     """
-    assert isinstance(operands, tuple)
-    return while_loop_op(cond_fn, body_fn, operands)
+    if torch._dynamo.is_compiling():
+        return while_loop_op(cond_fn, body_fn, operands)
+
```
Comment (zou3519) [inline torch/_higher_order_ops/while_loop.py:102]: We probalby want to validate the input even if we're compiling?

File: torch/_dynamo/variables/higher_order_ops.py [hunk 2]
```diff
@@ -734,6 +736,144 @@ def diff_meta(tensor_vars1, tensor_vars2):
         )
 
 
+class WhileLoopHigherOrderVariable(TorchHigherOrderOperatorVariable):
+    @raise_hard_error_if_graph_break(
+        reason="Cond doesn't work unless it is captured completely with torch.compile."
```
Comment (zou3519) [inline torch/_dynamo/variables/higher_order_ops.py:741]: while_loop

File: torch/_dynamo/variables/higher_order_ops.py [hunk 3]
```diff
@@ -734,6 +736,144 @@ def diff_meta(tensor_vars1, tensor_vars2):
...
+                (
+                    UserFunctionVariable,
+                    NestedUserFunctionVariable,
+                    NNModuleVariable,
+                    UnspecializedNNModuleVariable,
+                ),
+            ), str(type(fn_var))
+
+        _check_supported_callable(args[0])
+        _check_supported_callable(args[1])
```
Comment (zou3519) [inline torch/_dynamo/variables/higher_order_ops.py:780]: probably too strict (like cond -- I filed an issue)

File: torch/_dynamo/variables/higher_order_ops.py [hunk 4]
```diff
@@ -734,6 +736,144 @@ def diff_meta(tensor_vars1, tensor_vars2):
...
+
+        cond_r_meta = _extract_tensor_metadata(
+            cond_r.proxy.node.meta["example_value"], include_contiguity=False
+        )
+        if not cond_r_meta.dtype == torch.bool or not cond_r_meta.shape == torch.Size(
+            []
+        ):
+            unimplemented(
+                f"Expected cond_fn to return a tensor with shape (1,) but got {cond_r_meta.shape}"
+            )
```
Comment (zou3519) [inline torch/_dynamo/variables/higher_order_ops.py:815]: shape () ?

**Review by zou3519 — APPROVED** [review]
lgtm, please read comments

File: torch/_higher_order_ops/while_loop.py (hunk 1 above)
Comment (ydwu4) [reply torch/_higher_order_ops/while_loop.py:102]: I just tried this. The callable builit-in call is not supported in dynamo and we also have redundant check in WhileLoopHigherOrderOp. Can refactor it after supporting callable in dynamo. Basically, using your previous idea of inlining into python input checks in dynamo.

File: torch/_dynamo/variables/higher_order_ops.py (hunk 2 above)
Comment (ydwu4) [reply torch/_dynamo/variables/higher_order_ops.py:741]: Nice catch!

File: torch/_dynamo/variables/higher_order_ops.py (hunk 3 above)
Comment (ydwu4) [reply torch/_dynamo/variables/higher_order_ops.py:780]: yeah, I'll do these two together in next diff. Marked in my tasks list

**Pull request closed without merging**